staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -958,6 +999,12 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 				allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
 			} else {
 				for i, cr := range compResults {
+					expressionCost := getExpressionCost(cr.MaxCost, nodeCostInfo)
+					if expressionCost > ExpressionCostLimit {
+						exceedFactor := float64(expressionCost) / float64(ExpressionCostLimit)
+						costErrorMsg := fmt.Sprintf("CEL rule of cost %d exceeded budget of %d by factor of %vx", uint64(expressionCost), ExpressionCostLimit, exceedFactor)
```

Comment:
We might need to refine this error message a bit.  I tried out a case where there was a regex match under two lists, and there were no max limits set and the error I got was:

```
The CustomResourceDefinition "cel.stable.example.com" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[list].items.x-kubernetes-validations[0].rule: Forbidden: CEL rule of cost 109224385588891692 exceeded budget of 10000000 by factor of 1.092243855888917e+10x
```

The cost is crazy high (which is expected), and the factor doesn't make the information that much more digestible in this case.

I expect user to hit cases like this quite a bit because when they first start using CEL, they often will not have any previous reason to set max limits on their schemas.

A couple things to consider:
- round the factor to 1 decimal place and don't use an exponent, .e.g "%.1f"
- if the budged is exceeded by more than a factor that it helpful, don't print it, maybe 100x?
- Saying something to the user in the error about what they can do to fix the problem, .e.g "adding maxItems, maxProperties and maxLength validations to arrays, maps and strings may reduce this estimated cost"

Reply:
The current error message uses "%1.f" for displayed factors (and hides numbers if the factor is greater than 100), plus the recommendations for wording and solutions - I thought it might be good to not resolve just yet in case there were more ideas on making the message even more helpful.

Reply:
looks good to me